src/requestManager.cpp

```diff
@@ -984,52 +1020,56 @@ void CRequestManager::FindNextBlocksToDownload(NodeId nodeid, unsigned int count
     thindata.ClearThinBlockTimer(hash);
 
+    // Add to inflight, if it hasn't already been marked inflight for this node id.
     LOCK(cs_objDownloader);
-    std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
+    std::map<uint256, std::map<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
         mapBlocksInFlight.find(hash);
-    if (itInFlight == mapBlocksInFlight.end()) // If it hasn't already been marked inflight...
+    if (itInFlight == mapBlocksInFlight.end() || !mapBlocksInFlight[hash].count(nodeid))
     {
+        // Get a request manager nodestate pointer.
+        std::map<NodeId, CRequestManagerNodeState>::iterator it = mapRequestManagerNodeState.find(nodeid);
+        DbgAssert(it != mapRequestManagerNodeState.end(), return );
+        CRequestManagerNodeState *state = &it->second;
+
+        // Add queued block to nodestate and add iterator for queued block to mapBlocksInFlight
         int64_t nNow = GetTimeMicros();
         QueuedBlock newentry = {hash, nNow};
-        std::list<QueuedBlock>::iterator it = state->vBlocksInFlight.insert(state->vBlocksInFlight.end(), newentry);
+        std::list<QueuedBlock>::iterator it2 = state->vBlocksInFlight.insert(state->vBlocksInFlight.end(), newentry);
+        mapBlocksInFlight[hash][nodeid] = it2;
+
+        // Increment blocks in flight for this node and if applicable the time we started downloading.
         state->nBlocksInFlight++;
         if (state->nBlocksInFlight == 1)
         {
             // We're starting a block download (batch) from this peer.
             state->nDownloadingSince = GetTimeMicros();
         }
-        mapBlocksInFlight[hash] = std::make_pair(nodeid, it);
     }
 }
 
 // Returns a bool if successful in indicating we received this block.
 bool CRequestManager::MarkBlockAsReceived(const uint256 &hash, CNode *pnode)
 {
-    AssertLockHeld(cs_main);
+    if (!pnode)
+        return false;
 
     LOCK(cs_objDownloader);
-    std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
+    NodeId nodeid = pnode->GetId();
+    std::map<uint256, std::map<NodeId, std::list<QueuedBlock>::iterator> >::iterator itInFlight =
         mapBlocksInFlight.find(hash);
-    if (itInFlight != mapBlocksInFlight.end())
+    if (itInFlight != mapBlocksInFlight.end() && mapBlocksInFlight[hash].count(nodeid))
     {
-        CNodeState *state = State(itInFlight->second.first);
-        DbgAssert(state != nullptr, return false);
+        // Get a request manager nodestate pointer.
+        std::map<NodeId, CRequestManagerNodeState>::iterator it = mapRequestManagerNodeState.find(nodeid);
+        DbgAssert(it != mapRequestManagerNodeState.end(), return false);
+        CRequestManagerNodeState *state = &it->second;
 
-        int64_t getdataTime = itInFlight->second.second->nTime;
+        int64_t getdataTime = mapBlocksInFlight[hash][nodeid]->nTime;
```

Comment:
This double index must succeed because of the check you added on line 1065.  But doing it this way forces the code to lookup it up twice and does not make it completely clear that mapBlocksInFlight[hash][nodeid] is valid.  

Rather than checking via "count()", you might as well do a "find()" so that you get an iterator, test it for validity, and then use that iterator in this line.